kimchi/src/folding/expressions.rs

```diff
@@ -0,0 +1,304 @@
+    #[allow(dead_code)]
+    WitnessExtended(usize),
+    Index(GateType),
+    Coefficient(usize),
+    ///basically X, to allow accesing the next row
+    Shift,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+///A simplified expression to be use in folding
+pub enum FoldingExp<C> {
+    Constant(C),
+    Cell(FoldingColumn),
+    Double(Box<FoldingExp<C>>),
+    Square(Box<FoldingExp<C>>),
+    Add(Box<FoldingExp<C>>, Box<FoldingExp<C>>),
+    Sub(Box<FoldingExp<C>>, Box<FoldingExp<C>>),
+    Mul(Box<FoldingExp<C>>, Box<FoldingExp<C>>),
+    UnnormalizedLagrangeBasis(i32),
+    Cache(CacheId, Box<FoldingExp<C>>),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Degree {
+    Zero,
+    One,
+    Two,
+}
+impl<C> FoldingExp<C> {
+    pub(super) fn degree(&self) -> Degree {
+        use Degree::*;
+        match self {
+            FoldingExp::Constant(_) => Zero,
+            FoldingExp::Cell(col) => match col {
+                FoldingColumn::Witness(_) | FoldingColumn::WitnessExtended(_) => One,
+                _ => Zero,
+            },
+            FoldingExp::Double(e) => e.degree(),
+            FoldingExp::Square(e) => &e.degree() * &e.degree(),
+            FoldingExp::Add(e1, e2) | FoldingExp::Sub(e1, e2) | FoldingExp::Mul(e1, e2) => {
+                e1.degree() + e2.degree()
+            }
+            FoldingExp::UnnormalizedLagrangeBasis(_) => Zero,
+            FoldingExp::Cache(_, e) => e.degree(),
+        }
+    }
+}
+impl std::ops::Add for Degree {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        use Degree::*;
+        match (self, rhs) {
+            (Zero, Zero) => Zero,
+            (Zero, One) | (One, Zero) | (One, One) => One,
+            (_, Two) | (Two, _) => Two,
+        }
+    }
+}
+impl std::ops::Mul for &Degree {
+    type Output = Degree;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        use Degree::*;
+        match (self, rhs) {
+            (Zero, Zero) => Zero,
+            (Zero, One) | (One, Zero) => One,
+            (Zero, Two) | (One, One) | (Two, Zero) => Two,
+            (One, Two) | (Two, One) | (Two, Two) => panic!("degree over 2"),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Term<C> {
+    pub exp: FoldingExp<C>,
+    pub sign: bool,
+    pub degree: Degree,
+}
+
+impl<C> Term<C> {
+    fn new(exp: FoldingExp<C>, degree: Degree) -> Self {
+        Self {
+            exp,
+            sign: true,
+            degree,
+        }
+    }
+}
+impl<C> std::ops::Neg for Term<C> {
+    type Output = Self;
+
+    fn neg(self) -> Self::Output {
+        Term {
+            sign: !self.sign,
+            ..self
+        }
+    }
+}
+///A simplified expression with all terms separated by degree
+pub struct IntegratedFoldingExpr<C> {
+    pub(super) degree_0: Vec<(FoldingExp<C>, bool)>,
+    pub(super) degree_1: Vec<(FoldingExp<C>, bool)>,
```

Comment:
`degree_1` and `degree_2` are not used anywhere atm. I'd expect to see them in `compute_error` for cross term T computation, but there you only use `degree_0` -- I assume everything written passes tests, so I'm missing smth?